src/phase0/state_transition/epoch_processing.rs

```diff
@@ -444,6 +509,350 @@ pub fn process_participation_record_updates<
+        HISTORICAL_ROOTS_LIMIT,
+        ETH1_DATA_VOTES_BOUND,
+        VALIDATOR_REGISTRY_LIMIT,
+        EPOCHS_PER_HISTORICAL_VECTOR,
+        EPOCHS_PER_SLASHINGS_VECTOR,
+        MAX_VALIDATORS_PER_COMMITTEE,
+        PENDING_ATTESTATIONS_BOUND,
+    >,
+    context: &Context,
+) -> bool {
+    get_finality_delay(state, context) > context.min_epochs_to_inactivity_penalty
+}
+
+pub fn get_eligible_validator_indices<
+    const SLOTS_PER_HISTORICAL_ROOT: usize,
+    const HISTORICAL_ROOTS_LIMIT: usize,
+    const ETH1_DATA_VOTES_BOUND: usize,
+    const VALIDATOR_REGISTRY_LIMIT: usize,
+    const EPOCHS_PER_HISTORICAL_VECTOR: usize,
+    const EPOCHS_PER_SLASHINGS_VECTOR: usize,
+    const MAX_VALIDATORS_PER_COMMITTEE: usize,
+    const PENDING_ATTESTATIONS_BOUND: usize,
+>(
+    state: &BeaconState<
+        SLOTS_PER_HISTORICAL_ROOT,
+        HISTORICAL_ROOTS_LIMIT,
+        ETH1_DATA_VOTES_BOUND,
+        VALIDATOR_REGISTRY_LIMIT,
+        EPOCHS_PER_HISTORICAL_VECTOR,
+        EPOCHS_PER_SLASHINGS_VECTOR,
+        MAX_VALIDATORS_PER_COMMITTEE,
+        PENDING_ATTESTATIONS_BOUND,
+    >,
+    context: &Context,
+) -> Vec<ValidatorIndex> {
+    let previous_epoch = get_previous_epoch(state, context);
+    state
+        .validators
+        .iter()
+        .enumerate()
+        .filter_map(|(i, v)| {
+            if is_active_validator(v, previous_epoch)
+                || (v.slashed && previous_epoch + 1 < v.withdrawable_epoch)
+            {
+                Some(i)
+            } else {
+                None
+            }
+        })
+        .collect::<Vec<ValidatorIndex>>()
+}
+
+pub fn get_attestation_component_deltas<
+    const SLOTS_PER_HISTORICAL_ROOT: usize,
+    const HISTORICAL_ROOTS_LIMIT: usize,
+    const ETH1_DATA_VOTES_BOUND: usize,
+    const VALIDATOR_REGISTRY_LIMIT: usize,
+    const EPOCHS_PER_HISTORICAL_VECTOR: usize,
+    const EPOCHS_PER_SLASHINGS_VECTOR: usize,
+    const MAX_VALIDATORS_PER_COMMITTEE: usize,
+    const PENDING_ATTESTATIONS_BOUND: usize,
+>(
+    state: &BeaconState<
+        SLOTS_PER_HISTORICAL_ROOT,
+        HISTORICAL_ROOTS_LIMIT,
+        ETH1_DATA_VOTES_BOUND,
+        VALIDATOR_REGISTRY_LIMIT,
+        EPOCHS_PER_HISTORICAL_VECTOR,
+        EPOCHS_PER_SLASHINGS_VECTOR,
+        MAX_VALIDATORS_PER_COMMITTEE,
+        PENDING_ATTESTATIONS_BOUND,
+    >,
+    attestations: &[PendingAttestation<MAX_VALIDATORS_PER_COMMITTEE>],
+    context: &Context,
+) -> Result<(Vec<Gwei>, Vec<Gwei>), Error> {
+    // Helper with shared logic for use by get source, target, and head deltas functions
+    let mut rewards = vec![0, state.validators.len() as Gwei];
+    let mut penalties = vec![0, state.validators.len() as Gwei];
+    let total_balance = get_total_active_balance(state, context)?;
+    let unslashed_attesting_indices =
+        get_unslashed_attesting_indices(state, attestations, context)?;
+    let attesting_balance = get_total_balance(state, &unslashed_attesting_indices, context)?;
```

Comment:
This was the reason why I changed `get_total_balance` in `mod.rs`, so I'm open to suggestions if that's not desired. Basically, I was getting the error `borrow of moved value: unslashed_attesting_indices` -- which (checking my understanding) means `unslashed_attesting_indices` was being moved to `attesting_balance` when it calls `get_total_balance`, so the compiler can't access the original variable anymore. Thus, I couldn't access `unslashed_attesting_indices` within the `for` loop

Reply:
yeah, understanding is correct... i'll look at the change, prolly the right thing to do